flink-table/flink-table-common/src/main/java/org/apache/flink/table/factories/FileSystemFormatFactory.java

```diff
@@ -102,6 +105,66 @@
 		 * The follow up operator will filter the records again.
 		 */
 		List<Expression> getPushedDownFilters();
+
+		/**
+		 * Get field names without partition keys.
+		 */
+		default String[] getFieldNamesWithoutPartKeys() {
+			return Arrays.stream(getSchema().getFieldNames())
+				.filter(name -> !getPartitionKeys().contains(name))
+				.toArray(String[]::new);
+		}
+
+		/**
+		 * Get field types without partition keys.
+		 */
+		default DataType[] getFieldTypesWithoutPartKeys() {
+			return Arrays.stream(getSchema().getFieldNames())
+				.filter(name -> !getPartitionKeys().contains(name))
+				.map(name -> getSchema().getFieldDataType(name).get())
+				.toArray(DataType[]::new);
+		}
+
+		/**
+		 * RowType of table that excludes partition key fields.
+		 */
+		default RowType getNonPartRowType() {
+			return RowType.of(
+				Arrays.stream(getFieldTypesWithoutPartKeys())
+					.map(DataType::getLogicalType)
+					.toArray(LogicalType[]::new),
+				getFieldNamesWithoutPartKeys());
+		}
+
+		/**
+		 * Mapping from non-partition project fields index to all project fields index.
+		 */
+		default int[] getNonPartFieldProjectMapping() {
```

Comment:
These two methods should only one, otherwise it is very hard to understand.
Left only one: `getFormatProjectFields`